File: leetcode/110/src/P110.java
/**
 * Created by yfy on 2/25/16.
 * 110. Balanced Binary Tree
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public boolean isBalanced(TreeNode root) {
		return height(root) >= 0;
	}

	// Return -1 if not balanced
	private int height(TreeNode root) {
		if (root == null) return 0;
		int leftHeight = height(root.left);
		if (leftHeight < 0) return -1;
		int rightHeight = height(root.right);
		if (rightHeight < 0) return -1;
		if (Math.abs(leftHeight - rightHeight) <= 1)
			return Math.max(leftHeight, rightHeight) + 1;
		else
			return -1;
	}
}

public class P110 {
}

File: leetcode/110/P110.rb
# 110. Balanced Binary Tree
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
	height(root) >= 0
end

def height root
	return 0 if !root
	lh = height root.left
	return -1 if lh < 0
	rh = height root.right
	return -1 if rh < 0
	if (lh - rh).abs <= 1
		[lh, rh].max + 1
	else
		-1
	end
end

